### PR TITLE
Add link to TSan on macOS bug report.

### PR DESCRIPTION
**Pull request by tmcgilchrist**
This test case will hang under macOS x86_64 when run with `enable-tsan`.  Adding link back to the llvm bug report to save future developers chasing this down.

cc @OlivierNicole

### CODE REVIEW

**Comment by gasche** [comment]
Does the PR disable the test for the Tsan+macOS combination, or is there a reason not to do so?

**Comment by OlivierNicole** [comment]
Clearly correct and useful. I was wondering whether we should disable this test under macOS x86_64, but I’m not sure it’s easily doable without modify ocamltest (we can condition tests to the OS being macOS but not the architecture, IIRC?), and, since x86 Apple hardware is on a downward slope, maybe not worth the effort.

**Comment by OlivierNicole** [comment]
It looks like our comments were written in parallel :)

**Comment by gasche** [comment]
Ah, concurrency experts, they race to answer my questions in parallel with me asking them.

**Comment by tmcgilchrist** [comment]
We can depend on the architecture using `arch_amd64` from [builtin_actions.ml](https://github.com/ocaml/ocaml/blob/trunk/ocamltest/builtin_actions.ml#L368-L374), equivalent tests are available for other architectures. The expression to say not macOS on amd64 could be built into ocamltest as you said. It is not so easy to express that without a built in action.

**Comment by OlivierNicole** [comment]
I would say it’s up to you. I can’t take the time to add a builtin action right now, but i’m happy to review a PR that does. I would also be happy to merge this PR as-is.

**Review by OlivierNicole — APPROVED** [review]

**Comment by tmcgilchrist** [comment]
I'll write an adhoc action in ocamltest to cover it.

**Comment by shindere** [comment]
Tim McGilchrist (2024/06/18 07:36 -0700):
> I'll write an adhoc action in ocamltest to cover it.

For sake of transparency: we discussed it offline. :)

File: ocamltest/builtin_actions.ml [hunk 1]
```diff
@@ -148,6 +148,16 @@ let macos = make
     "on a MacOS system"
     "not on a MacOS system")
 
+let not_macos_amd64_tsan = make
+  ~name:"not_macos_amd64_tsan"
+  ~description:"Pass if not running on a MacOS amd64 system"
```
Comment (ghost) [inline ocamltest/builtin_actions.ml]:
```suggestion
  ~description:"Pass if not running on a MacOS amd64 system with TSan enabled"
```

File: ocamltest/builtin_actions.ml [hunk 2]
```diff
@@ -148,6 +148,16 @@ let macos = make
...
+  (Actions_helpers.pass_or_skip
+     (not ((Ocamltest_config.system = macos_system)
+           && (String.equal Ocamltest_config.arch "amd64")
+           && (Ocamltest_config.tsan)))
+     "not on a MacOS amd64 system"
+     "on a MacOS amd64 system")
```
Comment (ghost) [inline ocamltest/builtin_actions.ml]:
```suggestion
     "on a MacOS amd64 system with TSan enabled")
```

File: ocamltest/builtin_actions.ml [hunk 3]
```diff
@@ -148,6 +148,16 @@ let macos = make
...
     "not on a MacOS system")
 
+let not_macos_amd64_tsan = make
+  ~name:"not_macos_amd64_tsan"
+  ~description:"Pass if not running on a MacOS amd64 system"
+  (Actions_helpers.pass_or_skip
+     (not ((Ocamltest_config.system = macos_system)
+           && (String.equal Ocamltest_config.arch "amd64")
+           && (Ocamltest_config.tsan)))
+     "not on a MacOS amd64 system"
```
Comment (ghost) [inline ocamltest/builtin_actions.ml]:
```suggestion
     "not on a MacOS amd64 system, or TSan not enabled"
```

File: ocamltest/builtin_actions.ml (hunk 1 above)
Comment (tmcgilchrist) [reply ocamltest/builtin_actions.ml]: Quite right, I've included this correction in the latest squashed commit.

File: Changes [hunk 4]
```diff
@@ -394,6 +394,11 @@ _______________
   POWER.
   (Miod Vallat, review by Gabriel Scherer)
 
+- #13247: Disable lib_unix/kill test for MacOS AMD64 with TSan, linking
+  to llvm bug report causing infinite signal loops.
+  (Tim McGilchrist, review by OlivierNicole, Miod Vallat, Sébastien Hinderer,
```
Comment (ghost) [inline Changes]:
```suggestion
  (Tim McGilchrist, review by Olivier Nicole, Miod Vallat, Sébastien Hinderer,
```

**Review by ghost — APPROVED** [review]

File: testsuite/tests/lib-unix/kill/unix_kill.ml [hunk 5]
```diff
@@ -1,13 +1,17 @@
...
    bytecode;
  }{
    native;
  }
 *)
-
+(*
+  Possible infinite signal loop with TSan under MacOS
+  see https://github.com/llvm/llvm-project/issues/63824
+*)
```
Comment (OlivierNicole) [inline testsuite/tests/lib-unix/kill/unix_kill.ml]: Nit: I would have put the explanatory comment next to the line disabling the test under macOS+TSan.

**Review by OlivierNicole — APPROVED** [review]

File: testsuite/tests/lib-unix/kill/unix_kill.ml (hunk 5 above)
Comment (tmcgilchrist) [reply testsuite/tests/lib-unix/kill/unix_kill.ml]: Good idea, I didn't know ocamltest supported inline comments like this.

**Comment by shindere** [comment]
May I please ask for two further changes?

1. I'd suggest to put the inline comment in the test file above the line
it comments, rather than below it.

2. In `Changes`, I think the `and` is generally used in place of the
comma to separate the two last authors, not in addition to it.. So would
you mindremoving the comma before your and? And you could then also put
your and on the lone below so that the lines are not that crowded.

**Comment by tmcgilchrist** [comment]
@shindere I have included the two suggestions, please consider for merging if there are no further changes required. Thank you.

**Comment by OlivierNicole** [comment]
It looks to me like you have addressed the demand. I’m taking the liberty to merge, please blame me in case this is an issue.

**Comment by shindere** [comment]
Thanks both for having finished the work!

**Pull request merged**